### PR TITLE
Copy APK into a known location, so it can be easily discovered

### PR DESCRIPTION
**Pull request by blasten**
## Description

This allows the tool to discover the APK more easily, since the filename pattern is controlled by the plugin. The actual change in Dart can be done in https://github.com/flutter/flutter/pull/53369, which should now be as simple as reading from `$buildDir/app/outputs/flutter-apk/app-<abi>-<flavor-flag>-<build-mode-flag>.apk`.

## Related Issues

https://github.com/flutter/flutter/issues/52905

## Tests

I added the following tests: Updated the integration tests.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Nice!

File: packages/flutter_tools/gradle/flutter.gradle [hunk 1]
```diff
@@ -634,6 +634,38 @@ class FlutterPlugin implements Plugin<Project> {
...
+                    def outputDirectory = variant.packageApplicationProvider.get().outputDirectory
+                    //  `outputDirectory` is a `DirectoryProperty` in AGP 4.1.
+                    String outputDirectoryStr = outputDirectory.metaClass.respondsTo(outputDirectory, "get")
+                        ? outputDirectory.get() : outputDirectory
+                    String filename = "app"
+                    String abi = output.getFilter(OutputFile.ABI)
+                    if (abi != null && !abi.isEmpty()) {
+                        filename += "-${abi}"
+                    }
+                    if (variant.flavorName != null && !variant.flavorName.isEmpty()) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/gradle/flutter.gradle:656]: What does this do for multiple flavor dimensions? Do we get `dim1Dim2`?

Comment (blasten) [reply packages/flutter_tools/gradle/flutter.gradle:656]: Yup. I tested with multiple flavors and it works as expected

**Review by jonahwilliams — APPROVED** [review]
LGTM!

**Comment by anik786** [comment]
Thanks a lot 👍

**Pull request merged**